src/Expr.h

```diff
@@ -125,15 +125,24 @@ struct IRHandle : public IntrusivePtr<const IRNode> {
 
 /** Integer constants */
 struct IntImm : public ExprNode<IntImm> {
-    int value;
+    int64_t value;
 
-    static IntImm *make(int value) {
-        if (value >= -8 && value <= 8 &&
-            !small_int_cache[value + 8].ref_count.is_zero()) {
-            return &small_int_cache[value + 8];
+    static IntImm *make(Type t, int64_t value) {
+        internal_assert(t.is_int() && t.is_scalar()) << "IntImm must be a scalar Int\n";
+        internal_assert(t.bits == 8 || t.bits == 16 || t.bits == 32 || t.bits == 64)
+            << "IntImm must be 8, 16, 32, or 64-bit\n";
+
+        if (t.bits == 32 && value >= -8 && value <= 8 &&
+            !small_int_cache[(int)value + 8].ref_count.is_zero()) {
+            return &small_int_cache[(int)value + 8];
         }
+
         IntImm *node = new IntImm;
-        node->type = Int(32);
+        node->type = t;
+        // Normalize the value by dropping the high bits
```

Comment:
I assert it's in-range when coming in from user code via an implicit cast
(see for example operator+), but inside the compiler it's convenient to say
things like:

UIntImm::make(UInt(8), something + something_else)

or

UIntImm::make(UInt(8), some_uint64_t)

and have it handle the wrapping for you.

On Tue, Oct 6, 2015 at 11:03 AM, Zalman Stern <notifications@github.com>
wrote:

> In src/Expr.h
> <https://github.com/halide/Halide/pull/938#discussion_r41299675>:
>
> >          IntImm *node = new IntImm;
> > -        node->type = Int(32);
> > +        node->type = t;
> > +        // Normalize the value by dropping the high bits
>
> Is there a reason you chose to do this rather than to assert the value is
> in range?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/halide/Halide/pull/938/files#r41299675>.
>